sw/device/lib/dif/dif_pwrmgr.c

```diff
@@ -4,8 +4,563 @@
+  }
+
+  request_reg_info_t reg_info = request_reg_infos[req_type];
+  uint32_t reg_val = mmio_region_read32(pwrmgr->params.base_addr,
+                                        reg_info.current_sources_reg_offset);
+  *sources = bitfield_field32_read(reg_val, reg_info.bitfield);
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_request_sources_lock(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_req_type_t req_type) {
+  if (pwrmgr == NULL || !is_valid_req_type(req_type)) {
+    return kDifPwrmgrBadArg;
+  }
+
+  // Only a single bit of this register is significant, thus we don't perform a
+  // read-modify-write. Setting this bit to 0 locks sources.
+  mmio_region_write32(pwrmgr->params.base_addr,
+                      request_reg_infos[req_type].write_enable_reg_offset, 0);
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_request_sources_is_locked(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_req_type_t req_type,
+    bool *is_locked) {
+  if (pwrmgr == NULL || !is_valid_req_type(req_type) || is_locked == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  *is_locked = request_sources_is_locked(pwrmgr, req_type);
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_wakeup_request_recording_set_enabled(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t new_state) {
+  if (pwrmgr == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  bool enable = false;
+  if (!toggle_to_bool(new_state, &enable)) {
+    return kDifPwrmgrBadArg;
+  }
+
+  // Only a single bit of this register is significant, thus we don't perform a
+  // read-modify-write. Setting this bit to 1 disables recording.
+  uint32_t reg_val =
+      bitfield_bit32_write(0, PWRMGR_WAKE_INFO_CAPTURE_DIS_VAL, !enable);
+
+  mmio_region_write32(pwrmgr->params.base_addr,
+                      PWRMGR_WAKE_INFO_CAPTURE_DIS_REG_OFFSET, reg_val);
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_wakeup_request_recording_get_enabled(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t *cur_state) {
+  if (pwrmgr == NULL || cur_state == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  uint32_t reg_val = mmio_region_read32(
+      pwrmgr->params.base_addr, PWRMGR_WAKE_INFO_CAPTURE_DIS_REG_OFFSET);
+  // Recoding is disabled if this bit is set to 1.
+  *cur_state = bool_to_toggle(
+      !bitfield_bit32_read(reg_val, PWRMGR_WAKE_INFO_CAPTURE_DIS_VAL));
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_wakeup_reason_get(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_wakeup_reason_t *reason) {
+  if (pwrmgr == NULL || reason == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  uint32_t reg_val =
+      mmio_region_read32(pwrmgr->params.base_addr, PWRMGR_WAKE_INFO_REG_OFFSET);
+
+  dif_pwrmgr_wakeup_types_t types = 0;
+  if (bitfield_bit32_read(reg_val, PWRMGR_WAKE_INFO_FALL_THROUGH)) {
+    types |= kDifPwrmgrWakeupTypeFallThrough;
+  }
+  if (bitfield_bit32_read(reg_val, PWRMGR_WAKE_INFO_ABORT)) {
+    types |= kDifPwrmgrWakeupTypeAbort;
+  }
```

Comment:
You could make sure that `PWRMGR_WAKE_INFO_ABORT` and others match the public enum variants, and then you could get the whole wakeup-info register in one go, without if statements for every bit.

Up to you though.

Reply:
Good point, I chose this approach because these bits come after `REASONS` in the register layout so adding new sources would require the constants to be updated. We can revisit this after wakeup sources are stabilized. Moving these  before `REASONS` would also be nice. Captured in #3827.